src/group_theory/perm/list.lean

```diff
@@ -0,0 +1,354 @@
+begin
+  rcases l with (_|⟨x, _|⟨y, tl⟩⟩),
+  { simp },
+  { simp },
+  { simpa using form_perm_apply_head _ _ _ h }
+end
+
+lemma form_perm_eq_head_iff_eq_last (x y : α) :
+  form_perm (y :: l) x = y ↔ x = last (y :: l) (cons_ne_nil _ _) :=
+iff.trans (by rw form_perm_apply_last) (form_perm (y :: l)).injective.eq_iff
+
+lemma zip_with_swap_prod_support' (l l' : list α) :
+  {x | (zip_with swap l l').prod x ≠ x} ≤ l.to_finset ⊔ l'.to_finset :=
+begin
+  simp only [set.sup_eq_union, set.le_eq_subset],
+  induction l with y l hl generalizing l',
+  { simp },
+  { cases l' with z l',
+    { simp },
+    { intro x,
+      simp only [set.union_subset_iff, mem_cons_iff, zip_with_cons_cons, foldr, prod_cons,
+                 mul_apply],
+      intro hx,
+      by_cases h : x ∈ {x | (zip_with swap l l').prod x ≠ x},
+      { specialize hl l' h,
+        refine set.mem_union.elim hl (λ hm, _) (λ hm, _);
+        { simp only [finset.coe_insert, set.mem_insert_iff, finset.mem_coe, to_finset_cons,
+                     mem_to_finset] at hm ⊢,
+          simp [hm] } },
+      { simp only [not_not, set.mem_set_of_eq] at h,
+        simp only [h, set.mem_set_of_eq] at hx,
+        rw swap_apply_ne_self_iff at hx,
+        rcases hx with ⟨hyz, rfl|rfl⟩;
+        simp } } }
+end
+
+lemma zip_with_swap_prod_support [fintype α] (l l' : list α) :
+  (zip_with swap l l').prod.support ≤ l.to_finset ⊔ l'.to_finset :=
+begin
+  intros x hx,
+  have hx' : x ∈ {x | (zip_with swap l l').prod x ≠ x} := by simpa using hx,
+  simpa using zip_with_swap_prod_support' _ _ hx'
+end
+
+lemma support_form_perm_le' : {x | form_perm l x ≠ x} ≤ l.to_finset :=
+begin
+  refine (zip_with_swap_prod_support' l l.tail).trans _,
+  simpa [finset.subset_iff] using tail_subset l
+end
+
+lemma support_form_perm_le [fintype α] : support (form_perm l) ≤ l.to_finset :=
+begin
+  intros x hx,
+  have hx' : x ∈ {x | form_perm l x ≠ x} := by simpa using hx,
+  simpa using support_form_perm_le' _ hx'
+end
+
+lemma form_perm_apply_lt (xs : list α) (h : nodup xs) (n : ℕ) (hn : n + 1 < xs.length) :
+  form_perm xs (xs.nth_le n ((nat.lt_succ_self n).trans hn)) = xs.nth_le (n + 1) hn :=
+begin
+  induction n with n IH generalizing xs,
+  { simpa using form_perm_apply_nth_le_zero _ h _ },
+  { rcases xs with (_|⟨x, _|⟨y, l⟩⟩),
+    { simp },
+    { simp },
+    { specialize IH (y :: l) (nodup_of_nodup_cons h) _,
+      { simpa [nat.succ_lt_succ_iff] using hn },
+      simp only [swap_apply_eq_iff, coe_mul, form_perm_cons_cons, nth_le],
+      generalize_proofs at IH,
+      rw [IH, swap_apply_of_ne_of_ne, nth_le];
+      { rintro rfl,
+        simpa [nth_le_mem _ _ _] using h } } }
+end
+
+-- useful for rewrites
+lemma form_perm_apply_lt' (xs : list α) (h : nodup xs) (x : α) (n : ℕ) (hn : n + 1 < xs.length)
+  (hx : x = (xs.nth_le n ((nat.lt_succ_self n).trans hn))) :
+  (form_perm xs) x = xs.nth_le (n + 1) hn :=
+by { rw hx, exact form_perm_apply_lt _ h _ _ }
+
```

Comment:
Is this lemma used in this PR? If not, I'd be inclined to leave it out, then we can review whether it's actually useful in the follow-up PR that needs it.